### PR TITLE
Re-sync hsl()/hsla() implementation with the latest spec

### PR DESCRIPTION
**Pull request by weinig**
#### bc5b392973637f900bd2f3d89cfb722f08aaedf5
<pre>
Re-sync hsl()/hsla() implementation with the latest spec
<a href="https://bugs.webkit.org/show_bug.cgi?id=272941">https://bugs.webkit.org/show_bug.cgi?id=272941</a>

Reviewed by Darin Adler.

The CSS Color specification has changed since hsl()/hsla() was last
worked on. This updates things to support:

- Modified parsing rules which now allow &lt;number&gt; in places that previously
  supported &lt;percentage&gt;.
- Expanded value range for saturation and lightness, and a constricted value
  range for hue.
- Changes to the types used by the h/s/l/alpha symbols used in the relative
  color form to always be CSS_NUMBER.
- Update conversion to/from sRGB that accounts for expanded value range.

Since the value range is legal for values greater than what bounded sRGB
supports, the reference color space for HSLA was changed to ExpandedSRGB
and unnecessary clipping to the bounded form (such as after interpolation)
was removed.

* LayoutTests/TestExpectations:
    Two tests that use the expanded range of hsl() expecting it to be
    clipped now fail and will need to be revised with the working group.
    (Since the tests are reference tests, they now depend on the how the
    graphics library used by the engine maps out of gamut colors and the
    display properties it is being shown on).

* LayoutTests/TestExpectations
    Added two tests as new image only failures due to their dependence on
    an sRGB display and unspecified gamut mapping. Tracking with webkit.org/b/272940.

* LayoutTests/imported/w3c/web-platform-tests/css/css-color/parsing/color-computed-color-mix-function-expected.txt
* LayoutTests/imported/w3c/web-platform-tests/css/css-color/parsing/color-computed-relative-color-expected.txt
* LayoutTests/imported/w3c/web-platform-tests/css/css-color/parsing/color-invalid-relative-color-expected.txt
* LayoutTests/imported/w3c/web-platform-tests/css/css-color/parsing/color-valid-hsl-expected.txt
* LayoutTests/imported/w3c/web-platform-tests/css/css-color/parsing/color-valid-relative-color-expected.txt
* LayoutTests/imported/w3c/web-platform-tests/css/css-color/parsing/relative-color-out-of-gamut-expected.txt
    Update a good chunk of tests to now pass. We still see some hsl()/hsla() failures due to
    not implementing analogous component forwarding yet. Tracked via webkit.org/b/273033.

* LayoutTests/platform/glib/fast/css/hsl-color-expected.txt
* LayoutTests/platform/ios/fast/css/hsl-color-expected.txt
* LayoutTests/platform/mac/fast/css/hsl-color-expected.txt
* LayoutTests/platform/wincairo/fast/css/hsl-color-expected.txt
    This test had an out of gamut hsl() color which is now getting a slightly
    more green result. The render tree dump can&apos;t deal with out of gamut colors
    and we should probably fix that. Tracking with webkit.org/b/273032.

* Source/WebCore/css/parser/CSSPropertyParserConsumer+Color.cpp:
(WebCore::CSSPropertyParserHelpers::consumeRGBOrHSLLegacyOptionalAlphaRaw):
(WebCore::CSSPropertyParserHelpers::colorByResolvingHSLComponentsModern):
(WebCore::CSSPropertyParserHelpers::colorByResolvingHSLComponentsLegacy):
(WebCore::CSSPropertyParserHelpers::parseRelativeHSLParametersRaw):
(WebCore::CSSPropertyParserHelpers::parseNonRelativeHSLParametersLegacyRaw):
(WebCore::CSSPropertyParserHelpers::parseNonRelativeHSLParametersModernRaw):
(WebCore::CSSPropertyParserHelpers::parseNonRelativeHSLParametersRaw):
(WebCore::CSSPropertyParserHelpers::parseHSLParametersRaw):
(WebCore::CSSPropertyParserHelpers::parseColorFunctionRaw):
(WebCore::CSSPropertyParserHelpers::parseColorFunction):
(WebCore::CSSPropertyParserHelpers::colorByNormalizingHSLComponents): Deleted.
    Rework consumers to match the spec.

* Source/WebCore/css/parser/CSSPropertyParserConsumer+Color.h:
    Add option to clamp hsl()/hsla() components at parse time for temporary
    use by the canvas code.

* Source/WebCore/html/canvas/CanvasStyle.cpp:
    Opt into quirk hsl()/hsla() gamut clipping while the issue is being discussed
    in the working group. Tracking with webkit.org/b/272940.

* Source/WebCore/platform/graphics/ColorTypes.h:
    Update clamping behavior to preserve angles by using modular
    arithmetic rather than strict max/min clamping. This is needed
    now that angles are once again restricted to 0-360.
    Update reference color space for HSLA to ExtendedSRGBA.

* Source/WebCore/platform/graphics/ColorModels.h:
(WebCore::UnresolvedColorType::anyComponentIsNone const):
    Expose a anyComponentIsNone() helper, moved from Color.
    Update HSLA model to restrict angle and allow any saturation/lightness.

* Source/WebCore/platform/graphics/ColorNormalization.h:
    Remove automatic normalization of HSLA to 8-bit sRGB since we now allow
    out of gamut values.

* Source/WebCore/platform/graphics/Color.cpp:
(WebCore::Color::anyComponentIsNone const):
    Use the newly exposed anyComponentIsNone() function on UnresolvedColor.

* Source/WebCore/platform/graphics/ColorConversion.cpp:
(WebCore::ExtendedSRGBA&lt;float&gt;&gt;::convert):
(WebCore::HSLA&lt;float&gt;&gt;::convert):
* Source/WebCore/platform/graphics/ColorConversion.h:
    Update color conversion code to use update algorithm from the spec
    which allows for extended value ranges.

Canonical link: <a href="https://commits.webkit.org/277795@main">https://commits.webkit.org/277795@main</a>
</pre>

https://github.com/WebKit/WebKit/commit/6d60a9f4f7058797d3d6082ca2a787c87439c8b4

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [❌ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/48612 "1 style error") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/48/builds/27823 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/14/builds/51573 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/51299 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/32/builds/44677 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/50917 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/49/builds/33760 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/51/builds/25353 "Built successfully") | [  ~~🧪 wpe-wk2~~](https://ews-build.webkit.org/#/builders/34/builds/39753 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/49194 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/47/builds/25489 "Passed tests") | [  ~~🧪 api-mac~~](https://ews-build.webkit.org/#/builders/18/builds/41942 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/20850 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/42/builds/22967 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/10/builds/43119 "Passed tests") | [✅ 🛠 wpe-skia](https://ews-build.webkit.org/#/builders/52/builds/6668 "Built successfully") | 
| [✅ 🛠 🧪 jsc](https://ews-build.webkit.org/#/builders/20/builds/41911 "Built successfully and passed tests") | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/44905 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/36/builds/43605 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/53207 "Built successfully") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/44/builds/23659 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/50/builds/19957 "Passed tests") | [  ~~🧪 gtk-wk2~~](https://ews-build.webkit.org/#/builders/1/builds/47046 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/46/builds/24925 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/42147 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/45974 "Passed tests") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/10713 "Built successfully and passed tests") | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/43/builds/25729 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/45/builds/24646 "Built successfully") | | |

### CODE REVIEW

**Comment by weinig** [comment]
Looks like the following canvas tests are also assuming specific gamut mapping behavior (no surprise given their name):

imported/w3c/web-platform-tests/html/canvas/element/fill-and-stroke-styles/2d.fillStyle.parse.hsl-clamp-1.html
imported/w3c/web-platform-tests/html/canvas/element/fill-and-stroke-styles/2d.fillStyle.parse.hsla-clamp-1.html
imported/w3c/web-platform-tests/html/canvas/offscreen/fill-and-stroke-styles/2d.fillStyle.parse.hsl-clamp-1.html
imported/w3c/web-platform-tests/html/canvas/offscreen/fill-and-stroke-styles/2d.fillStyle.parse.hsl-clamp-1.worker.html
imported/w3c/web-platform-tests/html/canvas/offscreen/fill-and-stroke-styles/2d.fillStyle.parse.hsla-clamp-1.html
imported/w3c/web-platform-tests/html/canvas/offscreen/fill-and-stroke-styles/2d.fillStyle.parse.hsla-clamp-1.worker.html

Going to try turning off the CSS gamut mapping to see if that fixes the WPT/GTK bots.

**Comment by weinig** [comment]
Added a bunch of infrastructure to pass color parsing specific information down and used that to temporarily change the canvas color parsing code to continue to clamp while the test is still in flux.

**Comment by weinig** [comment]
Pulled a chunk out into https://github.com/WebKit/WebKit/pull/27547. Making a draft for now, will rebase on top once that lands.

**Comment by weinig** [comment]
Ready to go now that the other one is in.

File: LayoutTests/TestExpectations [hunk 1]
```diff
@@ -4707,6 +4707,10 @@ webkit.org/b/245347 imported/w3c/web-platform-tests/css/css-color/lch-010.html [
 webkit.org/b/245347 imported/w3c/web-platform-tests/css/css-color/oklch-009.html [ ImageOnlyFailure ]
 webkit.org/b/245347 imported/w3c/web-platform-tests/css/css-color/oklch-010.html [ ImageOnlyFailure ]
 
+# Tests assume hsl() gets clipped, which is no longer true.
```
Comment (darinadler) [inline LayoutTests/TestExpectations:4704]: Can we revise WPT as part of this same pull request?

File: Source/WebCore/css/parser/CSSPropertyParserConsumer+Color.cpp [hunk 2]
```diff
@@ -98,10 +100,25 @@ static Color consumeOriginColorRaw(CSSParserTokenRange& args, ColorParserState&
...
-static std::optional<double> consumeOptionalAlphaRaw(CSSParserTokenRange& range)
+static std::optional<double> consumeRGBOrHSLLegacyOptionalAlphaRaw(CSSParserTokenRange& args, double defaultValue = 1.0)
+{
+    if (!consumeCommaIncludingWhitespace(args))
+        return defaultValue;
+
+    if (auto alphaParameter = consumeNumberOrPercentRaw(args)) {
+        return WTF::switchOn(*alphaParameter,
+            [] (NumberRaw number) { return std::clamp(number.value, 0.0, 1.0); },
+            [] (PercentRaw percent) { return std::clamp(percent.value / 100.0, 0.0, 1.0); }
```
Comment (darinadler) [inline Source/WebCore/css/parser/CSSPropertyParserConsumer+Color.cpp]: Is `* 0.01` more efficient?

File: Source/WebCore/css/parser/CSSPropertyParserConsumer+Color.cpp [hunk 3]
```diff
@@ -359,125 +376,285 @@ template<RGBFunctionMode Mode> static Color parseRGBParametersRaw(CSSParserToken
...
     if (!saturation)
         return { };
 
-    auto lightness = consumePercentOrNoneRawAllowingSymbolTableIdent(args, symbolTable);
+    auto lightness = consumeNumberOrPercentOrNoneRawAllowingSymbolTableIdent(args, symbolTable);
     if (!lightness)
         return { };
 
-    auto alpha = consumeOptionalAlphaRawAllowingSymbolTableIdent(args, symbolTable);
+    // NOTE: In the relative syntax, the alpha value of the origin color is used if no alpha value
```
Comment (darinadler) [inline Source/WebCore/css/parser/CSSPropertyParserConsumer+Color.cpp]: I find this comment a little unclear. Is it explaining \why the code below does this? Can you reword to make that more obvious?

File: Source/WebCore/css/parser/CSSPropertyParserConsumer+Color.cpp [hunk 4]
```diff
@@ -359,125 +376,285 @@ template<RGBFunctionMode Mode> static Color parseRGBParametersRaw(CSSParserToken
...
+        // If any component is outside the reference range, we store the value as a HSLA<float> to allow for non-SRGB gamut values.
+        return hsla;
+    }
+
+    if (state.nestingLevel > 1) {
+        // If the color is being consumed as part of a composition (relative color, color-mix, light-dark, etc.), we store the value as a HSLA<float> to allow for maximum precision.
+        return hsla;
+    }
+
+    // NOTE: The explicit conversion to SRGBA<uint8_t> is an intentional performance optimization that allows storing the
```
Comment (darinadler) [inline Source/WebCore/css/parser/CSSPropertyParserConsumer+Color.cpp]: Not sure we need the word `NOTE`.

File: Source/WebCore/css/parser/CSSPropertyParserConsumer+Color.cpp [hunk 5]
```diff
@@ -359,125 +376,285 @@ template<RGBFunctionMode Mode> static Color parseRGBParametersRaw(CSSParserToken
...
```
Comment (darinadler) [inline Source/WebCore/css/parser/CSSPropertyParserConsumer+Color.cpp]: Not sure we need the word `NOTE`.

File: Source/WebCore/platform/graphics/Color.cpp [hunk 6]
```diff
@@ -183,8 +183,7 @@ bool Color::anyComponentIsNone() const
         if constexpr (std::is_same_v<ColorType, SRGBA<uint8_t>>) {
             return false;
         } else {
-            auto [c1, c2, c3, alpha] = underlyingColor.unresolved();
-            return std::isnan(c1) || std::isnan(c2) || std::isnan(c3) || std::isnan(alpha);
+            return underlyingColor.unresolved().anyComponentIsNone();
```
Comment (darinadler) [inline Source/WebCore/platform/graphics/Color.cpp:186]: Maybe take the braces out to match normal WebKit coding style.

File: Source/WebCore/platform/graphics/ColorConversion.cpp [hunk 7]
```diff
@@ -91,67 +91,65 @@ static HSLHueCalculationResult calculateHSLHue(float r, float g, float b)
...
-        return { grey, grey, grey, alpha };
-    }
+    auto [hue, saturation, lightness, alpha] = color.resolved();
 
-    // hueToRGB() wants hue in the 0-6 range.
-    auto scaledHue = (normalizeHue(hue) / 360.0f) * 6.0f;
-    auto scaledLightness = lightness / 100.0f;
-    auto scaledSaturation = saturation / 100.0f;
+    float normalizedHue = normalizeHue(hue);
+    float scaledSaturation = saturation / 100.0f;
```
Comment (darinadler) [inline Source/WebCore/platform/graphics/ColorConversion.cpp]: Same question about `* 0.01f`. More efficient?

**Review by darinadler — APPROVED** [review]

File: Source/WebCore/platform/graphics/Color.cpp (hunk 6 above)
Comment (nt1m) [reply Source/WebCore/platform/graphics/Color.cpp:186]: Could even remove the `else {` in this case to match WebKit style.

Comment (darinadler) [reply Source/WebCore/platform/graphics/Color.cpp:186]: In an `if constexpr` I think the `else` has meaning even after a `return`. So I suggest keeping it.

File: Source/WebCore/css/parser/CSSPropertyParserConsumer+Color.cpp (hunk 2 above)
Comment (weinig) [reply Source/WebCore/css/parser/CSSPropertyParserConsumer+Color.cpp]: Tested and multiplication does seem to be a tad faster (though it was much harder to get clang to cooperate and not compile things out than I remember, there must be a trick I forget). The following:

```c++
double division(double x)
{
    return x / 100.0;
}

double mutiplication(double x)
{
    return x * 0.01; /* or x * (1.0 / 100.0). Both produce identical assembly as one would expect */
}
```

compile to

```asm
__Z8divisiond:
0000000100003e9c	mov	x8, #0x4059000000000000
0000000100003ea0	fmov	d1, x8
0000000100003ea4	fdiv	d0, d0, d1
0000000100003ea8	ret
__Z13mutiplicationd:
0000000100003eac	adrp	x8, 0 ; 0x100003000
0000000100003eb0	ldr	d1, [x8, #0xf70]
0000000100003eb4	fmul	d0, d0, d1
0000000100003eb8	ret
```

under -0s.

We also use multiplication elsewhere in the file for normalization of rgb so I will update this to match.

File: Source/WebCore/css/parser/CSSPropertyParserConsumer+Color.cpp (hunk 4 above)
Comment (weinig) [reply Source/WebCore/css/parser/CSSPropertyParserConsumer+Color.cpp]: Agreed. Will remove.

File: LayoutTests/TestExpectations (hunk 1 above)
Comment (weinig) [reply LayoutTests/TestExpectations:4704]: It's been long enough a gap in contribution I totally forgot we could do that!

In this case though, I think waiting probably makes sense as I don't totally know what the right change to make is. I think the right outcome is likely just removing the tests (or at least renaming them, as they no longer are going to test clipping), but I want to check with editor first.

File: Source/WebCore/platform/graphics/ColorConversion.cpp (hunk 7 above)
Comment (weinig) [reply Source/WebCore/platform/graphics/ColorConversion.cpp]: Interestingly, replacing the `normalizedHue / 30.0f` below, with `normalizedHue * (1.0f / 30.0f)` leads to test failures due to rounding errors, so this is definitely a bit fiddly.

**Pull request merged**